### PR TITLE
[Impeller] Compute ContextContentOptions key via bit manipulating (instead of hashing each property).

### PR DESCRIPTION
**Pull request by jonahwilliams**
The hash function we use for pipeline hashncash is actually pretty slow. Since all of the properties we hash on fit in 64 bits, we can compute a hash value directly. From local benchmarking this is about twice as fast, which still leaves quite a bit of room for future improvement (faster hashmap? not using a hashmap at all?)

### CODE REVIEW

File: impeller/entity/contents/content_context.h [hunk 1]
```diff
@@ -279,11 +277,17 @@ struct ContentContextOptions {
...
-          o.has_stencil_attachment, o.wireframe, o.is_for_rrect_blur_clear);
+    constexpr uint64_t operator()(const ContentContextOptions& o) const {
+      return static_cast<uint64_t>(o.sample_count) << 56 |                   //
+             static_cast<uint64_t>(o.blend_mode) << 48 |                     //
+             static_cast<uint64_t>(o.stencil_compare) << 40 |                //
+             static_cast<uint64_t>(o.stencil_operation) << 32 |              //
+             static_cast<uint64_t>(o.primitive_type) << 24 |                 //
+             static_cast<uint64_t>(o.color_attachment_pixel_format) << 16 |  //
+             // bools
+             static_cast<uint64_t>(o.has_stencil_attachment) << 2 |
```
Comment (gaaclarke) [inline impeller/entity/contents/content_context.h:288]: This approach for the bools is correct.
```suggestion
             (o.has_stencil_attachment ? 1 : 0) << 2
```

File: impeller/entity/contents/content_context.h [hunk 2]
```diff
@@ -279,11 +277,17 @@ struct ContentContextOptions {
   bool is_for_rrect_blur_clear = false;
 
   struct Hash {
-    constexpr std::size_t operator()(const ContentContextOptions& o) const {
-      return fml::HashCombine(
-          o.sample_count, o.blend_mode, o.stencil_compare, o.stencil_operation,
-          o.primitive_type, o.color_attachment_pixel_format,
-          o.has_stencil_attachment, o.wireframe, o.is_for_rrect_blur_clear);
+    constexpr uint64_t operator()(const ContentContextOptions& o) const {
+      return static_cast<uint64_t>(o.sample_count) << 56 |                   //
```
Comment (gaaclarke) [inline impeller/entity/contents/content_context.h:281]: nit: I think this would be a bit easier to read if you count up instead of down.

Comment (gaaclarke) [inline impeller/entity/contents/content_context.h:281]: All of this is predicated on those fields being 8bits so you should add `static_assert`s here.

**Review by gaaclarke — COMMENTED** [review]
Nice find.  We should change the bool conversions and add static_asserts to be safe.

**Comment by gaaclarke** [comment]
Otherwise LGTM (fyi I'm OOO today so might have to approve tomorrow)

**Review by bdero — APPROVED** [review]
LGTM!

File: impeller/entity/contents/content_context.h (hunk 1 above)
Comment (jonahwilliams) [reply impeller/entity/contents/content_context.h]: Done

File: impeller/entity/contents/content_context.h (hunk 2 above)
Comment (jonahwilliams) [reply impeller/entity/contents/content_context.h]: Done

Comment (jonahwilliams) [reply impeller/entity/contents/content_context.h]: Done

File: impeller/entity/contents/content_context.h [hunk 3]
```diff
@@ -279,11 +277,25 @@ struct ContentContextOptions {
...
+    constexpr uint64_t operator()(const ContentContextOptions& o) const {
+      static_assert(sizeof(o.sample_count) == 1);
+      static_assert(sizeof(o.blend_mode) == 1);
+      static_assert(sizeof(o.sample_count) == 1);
+      static_assert(sizeof(o.stencil_compare) == 1);
+      static_assert(sizeof(o.stencil_operation) == 1);
+      static_assert(sizeof(o.primitive_type) == 1);
+      static_assert(sizeof(o.color_attachment_pixel_format) == 1);
+
+      return static_cast<uint64_t>(o.is_for_rrect_blur_clear ? 1u : 0u) << 0 |
```
Comment (gaaclarke) [inline impeller/entity/contents/content_context.h:289]: fyi you can get rid of these static casts if you use uint64_t literals.  I think its `1llu`?  something like that.

**Review by gaaclarke — APPROVED** [review]
lgtm!

File: impeller/entity/contents/content_context.h (hunk 3 above)
Comment (jonahwilliams) [reply impeller/entity/contents/content_context.h]: Done

**Pull request merged**